src/sms_verification/repository.rs

```diff
@@ -0,0 +1,208 @@
+            VerificationStatus::Verified => "VERIFIED",
+            VerificationStatus::Failed => "FAILED",
+        }
+    }
+}
+
+#[derive(Debug, Clone, sqlx::FromRow)]
+#[allow(dead_code)]
+pub struct SmsVerification {
+    pub unique_id: i32,
+    pub phone_number: String,
+    pub prelude_id: String,
+    pub created_at: DateTime<Utc>,
+    pub finalised_at: Option<DateTime<Utc>>,
+    pub signup_code: Option<Vec<u8>>,
+    pub status: VerificationStatus,
+    pub failure_reason: Option<String>,
+}
+
+#[derive(Clone, Debug)]
+pub struct SmsVerificationRepository {
+    db: SqlDb,
+}
+
+impl SmsVerificationRepository {
+    pub fn new(db: SqlDb) -> Self {
+        Self { db }
+    }
+
+    /// Create a new SMS verification record only if no pending session exists for this phone_number
+    pub async fn create_verification(
+        &self,
+        phone_number: &str,
+        prelude_id: &str,
+    ) -> Result<(), SmsVerificationRepositoryError> {
+        // Build subquery to check for existing pending sessions
+        let subquery = Query::select()
+            .expr(Expr::value(1))
+            .from("sms_verifications")
+            .and_where(Expr::col("phone_number").eq(phone_number))
+            .and_where(Expr::col("status").eq(VerificationStatus::Pending.as_str()))
+            .to_owned();
+
+        // Build INSERT statement with NOT EXISTS condition
+        let statement = Query::insert()
+            .into_table("sms_verifications")
+            .columns(["phone_number", "prelude_id"])
+            .select_from(
+                Query::select()
+                    .expr(Expr::value(phone_number))
+                    .expr(Expr::value(prelude_id))
+                    .cond_where(Expr::exists(subquery).not())
+                    .to_owned(),
+            )
+            .map_err(|e| DbError::QueryBuild(format!("Failed to build insert query: {}", e)))?
+            .to_owned();
+
+        let (query, values) = statement.build_sqlx(PostgresQueryBuilder);
+        sqlx::query_with(&query, values)
+            .execute(self.db.pool())
+            .await?;
+
+        Ok(())
+    }
+
+    pub async fn count_verified_sessions(
+        &self,
+        phone_number: &str,
+    ) -> Result<i64, SmsVerificationRepositoryError> {
+        let statement = Query::select()
+            .expr(Expr::col("unique_id").count())
+            .from("sms_verifications")
+            .and_where(Expr::col("phone_number").eq(phone_number))
+            .and_where(Expr::col("status").eq(VerificationStatus::Verified.as_str()))
+            .to_owned();
+        let (query, values) = statement.build_sqlx(PostgresQueryBuilder);
+        let row = sqlx::query_with(&query, values)
+            .fetch_one(self.db.pool())
+            .await?;
+        let count: i64 = row.try_get(0)?;
+        Ok(count)
+    }
+
+    /// Check if an active (pending) verification session exists for a phone number.
+    pub async fn check_pending_exists(
```

Comment:
Hmm, can we get a better name here? "Check" doesn't convey in which case this method errs.

```suggestion
    pub async fn err_if_no_active_verification()
```